### ISSUE TITLE
Ask about training.

### ISSUE THREAD

**Issue by 22nsuk** [lightvector/KataGo#176]
I'm using two GPUs to train weights. Most of the time is spent generating data. Can you give me a hint on how you conducted the training using a small GPUs?
I'm currently repeating data generation, shuffle, train and gatekeeping in order.
How did you conduct your training? If you were in sequence, how much data did you generate at a time? And how many cycles of training did you do with it?
If you didn't do it sequentially, I'd like to know what method you used.
I'd appreciate your reply.

**Comment by iopq** [comment]
It's being baked

https://github.com/lightvector/KataGo/pull/165

Right now I'm training on a 200,000 rows window, which in my case is around 12,000 games. I do an epoch with 1.0x learning rate, one with 0.1x learning rate, and one with 0.03x learning rate. Then I clean up all the games into another folder before starting again.

Right now I'm up to g104 level on 9x9

**Comment by lightvector** [comment]
As iopq noted, there hasn't really been enough testing to know for sure the best sequential methods. In particular, I'm still slightly uncertain about (1) - whether an oscillating learning rate schedule actually helps and especially (2) - holding the window sized fixed by "cleaning up" all the games every round, which would seem to promote overfitting and because in theory the window size should probably increase over time, not stay constant.

In case you're asking about the official run, the official run does everything asynchronously - all steps run simultaneously. Training and selfplay are balanced so that roughly there is anywhere from 4 to 20 times the number of training steps measured in number of samples ("s" number in the model name) as there is data generation measured in samples ("d" number in the model name). Since there are 8 symmetries, that would correspond to each orientation of a sample stochastically getting trained on somewhere from 0.5 to 2.5 times. The window size does grows dynamically over time, initially including just the last couple of nets or so and getting larger steadily.

I also used this method in early testing of KataGo when I only had 2 strong GPUs and one weak GPU. The two strong GPUs ran selfplay games, the weak GPU ran training, which achieved a ratio within the 4-20 above that I mentioned. All asynchronous, and it worked reasonably well, producing high-dan amateur nets on full 19x19 in merely a day or two. Once you have a working setup with gating (sometimes gating is helpful to be able to see if it's working for debugging) then you actually can just turn off gating too. Gating isn't really needed, so for a small-scale setup you can save the hassle and GPU power.

If you want to try asynchronous, you can perhaps cut the initial window size a little (adjust ` python/selfplay/shuffle.sh` and reduce min_rows from 250k to 100k or 50k) and the rest of the defaults should be reasonable as-is, because they are a all function of how many rows you generate, so if you generate fewer rows, the window size will grow slower as well.. And also perhaps cut the number of visits (adjust `cpp/configs/selfplay1.cfg` or `cpp/configs/selfplay2.cfg` to use fewer visits and fewer cheap search visits per move) to play more games a little faster. It can be possible to do asynchronous on just 2 GPUs as well without the extra weak GPU so long as your GPU has enough memory to do both selfplay and training. Tensorflow will normally hog more memory than it needs, so you can try `-gpu-memory-frac` for `train.py` or `train.sh` to see if you can push the number lower, and you can use `-sleep-seconds-per-epoch` to cause the training to run only intermittently, so as to achieve the right balance between rate of selfplay and training steps, and have one pure-selfplay machine, one selfplay machine that intermittently also runs training concurrently, and no gating.

**Comment by iopq** [comment]
the maximum rows should be cut too, there's no reason why you should be holding very large numbers (millions) at the beginning of a run

**Comment by lightvector** [comment]
@iopq - If you are referring to `-keep-target-rows` then yes good point, you can also cut that. It will save some CPU and disk space. Indeed, if you have the shuffle script running in a loop, you can just set that number so that it's a little larger than the number of training samples that you consume in the same amount of time. It doesn't actually matter *right* at the start though, when the window is already even smaller than that though.

A little point of clarification: Right now all scripts in the repo are "designed" for asynchronous by default, although in the last month or two there have been minor extra flags and options added to allow synchronous to be possible. Every script and process is built to be capable of running continuously, polling the appropriate directory for new data and outputting data for the next step (selfplay polls for new models and writes data, shuffle loop takes data and writes shuffled data, training loop takes latest shuffled data and saves new model in Tensorflow format, export loop polls for new models in Tensorflow format and converts to KataGo model format for selfplay).

I may consider to test and add some more "official" helper scripts to directly support synchronous in the future though. But at least the first version will perhaps not have cyclic learning rates and window clearing.  :)

**Comment by iopq** [comment]
Cyclic learning rates are commented out

**Comment by 22nsuk** [comment]
Thank you for your kind and detailed explanation. Here's another question.
First, you said that theoretically, the size of the window should get bigger. What is the reason?
The second question is the rate of learning. How do you calculate the timing to properly reduce your learning rate? 
Thank you always for doing good for the community.

**Comment by lightvector** [comment]
If the window size is too small, then adjacent samples are too correlated and cause overfitting. In the extreme case of window size of just a few games, you overfit to those games, by training on positions from them dozens of times, then the next few games of dozen times, etc. Or if the window size is only as big as about 1 previous net's total selfplay games, then you may overfit to the moves and evaluations of just that net, and then for the next net, overfit to that one, etc. Any fixed window size contributes some factor on how much "overfitting bias" you get of this sort, and early on it doesn't matter because you're learning so many big important things anyways that some small bias matters less, but as you converge, it seems you should also want to decrease this gradually.

A second separate reason is that the cost of a large window size drops greatly over time. Early on, nets improve fast, and so even slightly older nets are much weaker and produce much lower-quality selfplay games. But the longer you go, the slower the nets improve, and that means big window size costs  less and less because the older nets in your window are closer and closer to being about the same strength as the most recent nets. Since bigger window size costs less and less, the optimal window size should therefore correspondingly grow.

Learning rate - I just used a constant learning rate for almost everything. KataGo used `-lr-scale 1.0` (so, just the default learning rate scaling) all the way from random play to nearly as strong as g104, i.e. already at the strength of ELFv2 on 19x19. After that we started to try to drop the learning rate, very very gradually. More-fancy scaling methods might help, but just using constant learning rate almost forever already works very well, and is super-simple: nothing to calculate or tune.

My mental model for learning rate is that for a particular network size, any fixed learning rate has a ceiling that it will not easily let the net improve beyond due to the noise from that level of learning rate. And short of this ceiling, higher learning rate is typically better. Any time you lower the learning rate, you will observe a one-time boost in the strength due to reducing noise, but any subsequent learning will be slower, and if your window size is also small, then you also risk too-finely overfitting to the data in your small window (this is my primary concern with @iopq's method). Therefore you usually want learning rate high as long as possible, dropping it only when you decide you are not going to do much training any more (to get the one-time boost) or because you've hit the ceiling for that LR.

But it seems for 20 blocks x 256 channels blocks this ceiling is at or beyond ELFv2 strength already, so just using a constant high learning rate for a very long time works fine.

**Comment by 22nsuk** [comment]
Thank you. Your answer was very helpful.
